python/ray/data/impl/plan.py

```diff
@@ -0,0 +1,153 @@
+from ray.data.block import Block
+from ray.data.impl.block_list import BlockList
+from ray.data.impl.compute import get_compute
+from ray.data.impl.stats import DatasetStats
+
+from typing import Callable, Tuple, Optional, Union, TYPE_CHECKING
+import uuid
+
+if TYPE_CHECKING:
+    import pyarrow
+
+
+class ExecutionPlan:
+    def __init__(self, in_blocks: BlockList, stats: DatasetStats):
+        self._in_blocks = in_blocks
+        self._out_blocks = None
+        self._in_stats = stats
+        self._out_stats = None
+        self._stages = []
+        self._dataset_uuid = uuid.uuid4().hex
+        if not stats.dataset_uuid:
+            stats.dataset_uuid = self._dataset_uuid
+
+    def with_stage(self, stage: "Stage"):
+        if self._out_blocks:
+            copy = ExecutionPlan(self._out_blocks, self._out_stats)
+            copy._stages = [stage]
+        else:
+            copy = ExecutionPlan(self._in_blocks, self._in_stats)
+            copy._stages = self._stages.copy()
+            copy._stages.append(stage)
+        return copy
+
+    def initial_num_blocks(self) -> int:
+        if self._out_blocks:
+            return self._out_blocks.initial_num_blocks()
+        for stage in self._stages[::-1]:
+            if stage.num_blocks is not None:
+                return stage.num_blocks
+        return self._in_blocks.initial_num_blocks()
+
+    def schema(
+        self, fetch_if_missing: bool = False
+    ) -> Union[type, "pyarrow.lib.Schema"]:
+        if self._stages:
+            if fetch_if_missing:
+                self.execute()
+            blocks = self._out_blocks
+        else:
+            blocks = self._in_blocks
+        metadata = blocks.get_metadata() if blocks else []
+        # Some blocks could be empty, in which case we cannot get their schema.
+        # TODO(ekl) validate schema is the same across different blocks.
+        for m in metadata:
+            if m.schema is not None:
+                return m.schema
+        if not fetch_if_missing:
+            return None
+        # Need to synchronously fetch schema.
+        return blocks.ensure_schema_for_first_block()
+
+    def meta_count(self) -> Optional[int]:
+        if self._stages:
+            blocks = self._out_blocks
+        else:
+            blocks = self._in_blocks
+        metadata = blocks.get_metadata() if blocks else None
+        if metadata and metadata[0].num_rows is not None:
+            return sum(m.num_rows for m in metadata)
+        else:
+            return None
+
+    def execute(self, clear_input_blocks: bool = True) -> BlockList:
+        # TODO: add optimizations:
+        # 1. task fusion of OneToOne
+        # 2. task fusion of OneToOne to AlltoAll
+        # 3. clear input blocks
+        if self._out_blocks is None:
+            blocks = self._in_blocks
+            stats = self._in_stats
+            for stage in self._stages:
```

Comment:
Assuming that we'll want to support lazy fan-in operations `(Dataset, ..., Dataset) --> Dataset` in a future PR, such as union, zip, join, etc., do you have an idea of how this would be tweaked to support more than the currently supported chains of 1-to-1 block list transformations?

It seems as if you'd need to support something like external references to execution plans for other `Dataset` objects, and allow the `Dataset` that's consumed (the one that's the output of the e.g. join) to trigger execution of those other execution plans. It might be easier/cleaner to change the execution walk to bottom-up recursive rather than top-down iterative, at that point.

Reply:
I think this can be added in a future date, given the priority is to optimize ingest pipelines.